Use cockpit's PatternFly CSS

martinpitt opened the pull request:
Cockpit's patternfly.css already sets the correct font paths, and does
not have any other modifications right now (Cockpit specific
modifications live in the separate cockpit.min.css).

This will automatically pick up future adjustments to Cockpit PatternFly
styling, such as in https://github.com/cockpit-project/cockpit/pull/11987,
nd thus ensure that all Cockpit pages will have a consistent look and
feel.

This also avoids shipping the PatternFly css and fonts and thus makes
the package much smaller.

 - [x] Land cockpit PF4 update: https://github.com/cockpit-project/cockpit/pull/11987
 - [x] Fix kebabs and other UI elements in Cockpit's PF: https://github.com/cockpit-project/cockpit/pull/12211
 - [ ] Test all pages and add composer-local fixes

martinpitt commented:
When running against currently released cockpit, this makes no difference. But when running against the [PF4 styling branch](https://github.com/cockpit-project/cockpit/pull/11987#issuecomment-501663275), it will automatically adopt to the new style, and at first sight it looks just fine:

![bplist](https://user-images.githubusercontent.com/200109/59487585-79618080-8e7d-11e9-85e4-8686d31ee732.png)

![bpedit](https://user-images.githubusercontent.com/200109/59487588-7bc3da80-8e7d-11e9-9c09-efad2a4dbb9c.png)

![bpuser](https://user-images.githubusercontent.com/200109/59487592-7ebecb00-8e7d-11e9-9f26-d4e8ef9feeb9.png)

martinpitt commented:
So this is at least useful to explore Composer's UI with a PF4 look and feel, and discover CSS bugs.

I still have a bad gut feeling about this, though -- it seems to me that long-term this is more brittle than necessary. I don't yet know how this will behave with PF4 components once we start migrating them one by one -- if they still require a global patternfly.css `<link>` or whether this will be handled by local CSS via webpack.

So long-term I think we should copy cockpit's PF CSS adjustments and build them ourselves in Composer. I'll work on such a branch next.

@larskarlitski , @jgiardino , do you have an opinion about this?

martinpitt commented:
@garrett: FYI

martinpitt commented:
On second thought, cockpit-composer just takes the verbatim ./node_modules/patternfly/dist/css/patternfly.min.css and ships it as-is. Cockpit does the same plus:

 - [adjusting font paths](https://github.com/cockpit-project/cockpit/blob/master/tools/patternfly.sed) (which is fine for Composer),
 - and soon, the the PF4 look (which we want for composer)

and nothing else. All other customizations are *not* shipped via patternfly.css, but via cockpit.min.css.

Also, this patternfly.css will always stay PF3 (PF4 has its own independent files), so what we do about PF4 widgets is not influenced by the decision here.

After these investigations, I retract my claim that this is more brittle, and it's actually fairly elegant. It even avoids shipping a duplicate patternfly.css and the fonts, i. e. makes cockpit-composer smaller.

jgiardino commented:
Thanks for doing this, @martinpitt! 

Just looking at the examples you posted, there are a few issues I notice. I'm guessing these are related to either custom css that I included or pf3 css that I am using that cockpit is not. I don't imagine these issues would be difficult to fix, and I'm happy to work on these.

Would it make sense for me to pull these changes and make further updates for these issues as part of this PR?

martinpitt commented:
@jgiardino : by "examples I posted" you mean the screenshots from the PF4 look? Indeed there are still a lot of little bugs with that, see the long TODO list in https://github.com/cockpit-project/cockpit/pull/11987. These screenshots merely demonstrated that with this PR c-composer will adjust to cockpit's patternfly.css. However, the PF4 look will still take some time to land, and part of it will be to also validate composer.

This PR by itself should not change anything in the UI really. Did you notice any visual regressions there? I didn't. This PR would merely pave the way so that we *can* move to the PF4 look.

So we mostly need to decide whether we conceptually want to share Cockpit's PF CSS with cockpit-{ostree,podman,composer} (the former two already do), or whether cockpit-composer wants to retain full control of its own PatternFly, and then update to the PF4 look in lockstep.

martinpitt commented:
As far as I understood @larskarlitski and @jgiardino , we want to go with this approach, but postpone it until after the cockpit changes are done and land. Then we can add any possibly required fixups here, and don't risk having an intermediate broken layout. Did I get this right?

Marking as blocked, and added a TODO  list.

larskarlitski commented:
> As far as I understood @larskarlitski and @jgiardino , we want to go with this approach, but postpone it until after the cockpit changes are done and land. Then we can add any possibly required fixups here, and don't risk having an intermediate broken layout. Did I get this right?

Indeed. Thanks!

martinpitt commented:
Unblocking, as the PF4 PR landed now. Force-pushed to fix the trivial conflict.

I can't build cockpit-composer right now due to dozens of errors like
```
/home/martin/upstream/cockpit-composer/components/ListView/ComponentDetailsView.js
   76:19   error  'component.builds' is missing in props validation                    react/prop-types
```
I'll investigate that now.

martinpitt commented:
I was finally able to build locally with PR #710. I updated the screenshots above. For sure the kebab vs. button alignment/spacing/frames need work, but at least it's usable. Someone with a more trained "pixel error" eye than mine now needs to go through with a finer comb.

garrett commented:
Oh, it's looking pretty good! Yeah, the kebab button needs work. I know where to change it and how to change it; I should do that.

There are some widgets (like kebab) that are used in Composer but not in Cockpit; these would need to be added to the overrides file (in this case, integrated into the existing button/dropdown CSS). There are probably other widgets and groups of widgets which might also not be covered yet too.

I need to get Composer working locally. :wink:

martinpitt commented:
> I need to get Composer working locally.

I just did `dnf install lorax-composer`, and then I can just run cockpit-composer out of the build tree. What breaks for you?

So you are saying we should fix the kebabs straight in Cockpit's PF overrides, not in composer's CSS? That makes sense indeed. Thanks for looking into it!

garrett commented:
@martinpitt: When I try to run Composer, this is what I see:

![Screenshot_2019-07-03 Image Builder - Sunny](https://user-images.githubusercontent.com/10246/60572686-6a803680-9d76-11e9-9138-ba22aea6ccd7.png)

As to not have a bunch of noise (from extensions and whatnot), I also attempted this in Epiphany and checked the console there. This is what I see:

![composer-error](https://user-images.githubusercontent.com/10246/60573020-16c21d00-9d77-11e9-9e7c-9852597afb31.png)

garrett commented:
Here's the console from Chrome:

![chrome-console](https://user-images.githubusercontent.com/10246/60573999-fdba6b80-9d78-11e9-925a-56a354e7903b.png)

jgiardino commented:
Thanks for making these updates, @martinpitt! They look amazing, @garrett! Aside from the kebab, these are some other things I noticed in the screencaps above (but still need to review across all the pages):

- List view icon circle isn't sized right relative to the icon. This is based on the [PF3 list view](https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-rows.html)
  ![image](https://user-images.githubusercontent.com/21063328/60597805-93ee9180-9d79-11e9-9e18-dd9ba9322a88.png)

- On the edit blueprint page, in the Blueprint Components, the select menu before the filter input is behind the filter input:
  ![image](https://user-images.githubusercontent.com/21063328/60597745-73bed280-9d79-11e9-8866-32acee9e2a4c.png)
  and what it looks like in PF3:
  ![image](https://user-images.githubusercontent.com/21063328/60597871-b41e5080-9d79-11e9-9ecd-316dc625b594.png)

martinpitt commented:
Force-pushed to pick up the empty state error message fix, so that this branch doesn't have regressions.

garrett commented:
I've fixed most of the above issues — and several other small ones that affect Composer — in https://github.com/cockpit-project/cockpit/pull/12211.

The notable exception is the list view icon circle. I'll have to address it next week in another round of fixups. (There are a few other small ones I ran out of time to address this week.)

martinpitt commented:
I updated the screenshots for https://github.com/cockpit-project/cockpit/pull/12211 . I'll merge this later tonight, and then build updated RPMs for @jgiardino . Thanks @garrett!

garrett commented:
I just noticed there's a bug with the dropdowns when they have a selected state. I'll have to fix next week. (On Monday.) It's a very simple fix, but it's EOB/EOW for me now. We'll get it fixed before release. :wink:

martinpitt commented:
@jgiardino : I updated the RPMs in https://piware.de/tmp/cpf4/ to include the latest bug fixes, if you want to play around with it.

martinpitt commented:
@jgiardino : I updated the RPMs in https://piware.de/tmp/cpf4/ again with the latest of @garrett's fixes. Do you have some time today to test this? It would be nice to do a composer release with this tomorrow, so that it updates in lockstep with cockpit in Fedora 30.

martinpitt commented:
@jgiardino : You can test these in your RHEL 8.1 VM with:
```
wget -r -np https://piware.de/tmp/cpf4/
rpm -F --verbose piware.de/tmp/cpf4/*.rpm
systemctl stop cockpit
```

After that, `rpm -qa cockpit*` should show all 197.x-* versions.

martinpitt commented:
Copying @jgiardino's comment from PR #725 here for having a full record here. @garrett, these may be interesting to apply to Cockpit itself as well?

[I] add[ed] a few miscellaneous css updates in the context of Composer custom css updates.

There were a couple of items provided in the cockpit css that didn't have sufficient color contrast. I updated those here, but it might be worth checking if these are used in cockpit, and if the color contrast issues are present there as well.
```
.fields-status-pf { 	
  color: #8b8d8f;
  margin-bottom: 15px;
}
```
 I updated this in composer to `var(--pf-global--Color--200);`

```
.label-info { 
  background-color: #73bcf7;
}
```
 I updated this in composer to `var(--pf-global--link--Color)`

jgiardino commented:
Thanks, @martinpitt!

![bitmoji](https://render.bitstrips.com/v2/cpanel/f70171d9-4f76-4409-b78f-8ed9b0076e12-fe7ba5ad-9c01-4dac-b094-998c3f204a3b-v1.png?transparent=1&palette=1&width=246)
